crates/astria-celestia-client/src/client.rs

```diff
@@ -0,0 +1,306 @@
+}
+
+pub struct GetRollupDataResponse {
+    pub height: u64,
+    pub namespace: Namespace,
+    pub datas: Vec<SignedNamespaceData<RollupNamespaceData>>,
+    pub bad_blobs: Vec<BadBlob>,
+}
+
+#[async_trait]
+pub trait CelestiaClientExt: BlobClient {
+    /// Fetch sequencer data at the given height with the provided namespace.
+    ///
+    /// Returns successfully deserialized blobs in the `.datas` field. The
+    /// `.bad_blobs` field contains the celestia commitment for each blob
+    /// that could not be turned into sequencer data and the reason for it.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the underlying `blob.GetAll` JSONRPC failed.
+    async fn get_sequencer_data(
+        &self,
+        height: u64,
+        namespace: Namespace,
+    ) -> Result<GetSequencerDataResponse, jsonrpsee::core::Error> {
+        let blobs = self.blob_get_all(height, &[namespace]).await?;
+
+        let mut datas = Vec::new();
+        let mut bad_blobs = Vec::new();
+        for blob in blobs {
+            if blob.namespace != namespace {
+                bad_blobs.push(BadBlob {
+                    reason: BadBlobReason::WrongNamespace(blob.namespace),
+                    commitment: blob.commitment,
+                });
+            }
+            match serde_json::from_slice(&blob.data) {
+                Ok(data) => datas.push(data),
+                Err(err) => bad_blobs.push(BadBlob {
+                    reason: BadBlobReason::Deserialization(err),
+                    commitment: blob.commitment,
+                }),
+            }
+        }
+
+        Ok(GetSequencerDataResponse {
+            height,
+            namespace,
+            datas,
+            bad_blobs,
+        })
+    }
+
+    /// Returns the rollup data for a given rollup namespace at a given height, if it exists.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// + the verification key could not be constructed from the data stored in `namespace_data`;
+    /// + the RPC to fetch the blobs failed.
+    async fn get_rollup_data_matching_sequencer_data(
+        &self,
+        height: u64,
+        namespace: Namespace,
+        sequencer_data: &SignedNamespaceData<SequencerNamespaceData>,
+    ) -> Result<Vec<RollupNamespaceData>, jsonrpsee::core::Error> {
+        #[must_use]
+        fn is_blob_not_found(err: &jsonrpsee::core::Error) -> bool {
+            if let jsonrpsee::core::Error::Call(err) = err {
+                return err.message().contains("blob: not found");
+            }
+            false
+        }
+
+        let rsp = self.blob_get_all(height, &[namespace]).await;
+        let blobs = match rsp {
+            Ok(blobs) => blobs,
+            Err(err) if is_blob_not_found(&err) => {
+                return Ok(vec![]);
+            }
+            Err(err) => {
+                return Err(err);
+            }
+        };
+
+        // filter out blobs; we should only be left with either zero or one rollup datas
```

Comment:
remove/update comment, i assume if multiple rollup datas are returned the caller needs to handle it now?

Reply:
good catch! yeah, this should be handled by the caller now.